apis/experimental/node_classes/types.go

```diff
@@ -0,0 +1,152 @@
+	metav1.ObjectMeta
+
+	// ...
+}
+
+// This is the level that cloud-agnostic controllers would interact with. This
+// is where one could set the desired size/shape of a NodePool.
+type NodePool struct {
+	metav1.ObjectMeta
+
+	Spec NodePoolSpec
+	Status NodePoolStatus
+}
+
+type NodePoolSpec struct {
+	Name string // foo
+	Class string // name of NodeClass
+	Args string // JSON { ... } or ugly substruct with cloud-specific fields.
+	            // this is explained in NodeClass.
+	Size uint64
+	MinSize uint64
+	MaxSize uint64
+}
+
+type NodePoolStatus struct {
+	Phase NodePoolPhase
+	Message string
+	Reason string
+
+	// In addition to phase, we probably want a representation of in-flight
+	// cloud requests so we don't duplicate them in further runs of the
+	// controller loop. Once we decide to spin up a new node, we should
+	// make note of that because it may still take a few minutes for it to
+	// become healthy and register, and we don't want to keep triggering
+	// instance creation just because current < desired. This could be
+	// avoided if we always rely on an ASG/MIG, but that might not give us
+	// the control we want.
+	PendingNodeAdditions []string
+	PendingNodeDeletions []string
+}
+
+type NodePoolPhase string
+const (
+	NodePoolPending NodePoolPhase = "Pending"
+	NodePoolRunning NodePoolPhase = "Running"
+	NodePoolTerminated NodePoolPhase = "Terminated"
+	// ...
+)
+
+// These are objects that a cluster admin (or infrastructure admin, if a
+// differnet role) would create in order to set the policy of what kinds of
+// nodes are available and allowed for the cluster, analogous to
+// PersistentVolumes and PersistentVolumeClaims. They can have generic names
+// like "small," "medium," "gpu," or as arbitrarily complex as is useful
+// ("large-gpu-us-west-ssd"). In a cloud context, it allows infrastructure
+// admins to place restrictions on what resources are allowed to be created,
+// and in an on-premise context, allows accurate modeling of what hardware is
+// actually available for use.
+//
+// These serve as the glue between NodePools and NodeBuilders. In order to keep
+// the NodeClass registry from becoming a combinatorial explosion, we can
+// parameterize attributes that aren't core to the NodeClass' identity. For
+// instance, it might be very important that region/zone is a core part of a
+// NodeClass, and an admin wants their naming scheme to always start with the
+// zone. However, within that zone, the exact disk image you use is far more
+// flexible and inconsequential to the cluster admin, so it's modeled as an
+// overridable attribute. NodePools can then specify Args to fill in the blanks
+// or override defaults, and NodeBuilders will merge the Args into the Params
+// and use the result to populate a node's attributes.
+type NodeClass struct {
+	Name string
+	Params NodeClassParams
+	NodeBuilder api.ObjectReference
+}
+
+type NodeClassParams struct {
+	InstanceType string // n1-standard-1, r2.medium, etc.
```

Comment:
Do you want this to be cloud specific? The PV interface is a little more generic, last I looked. It seems more likely you'd want to e.g. specify a core count (and e.g. GPU count, or whether the hardware is dedicated (possible on some clouds), preemptible, etc.), though comparing these across cloud providers is perilous.

Reply:
Ah, good point. This is definitely an area that could use more exploration and prototyping to see what actually makes sense.